src/index.ts

```diff
@@ -44,21 +46,109 @@ interface StorageMetadata extends Record<string, string> {
 	}
 
 	const buffer = await request.arrayBuffer();
-	const tokenRequest = TokenRequest.deserialize(new Uint8Array(buffer));
+	const tokenRequest = TokenRequest.deserialize(TOKEN_TYPES.BLIND_RSA, new Uint8Array(buffer));
 
 	if (tokenRequest.tokenType !== TOKEN_TYPES.BLIND_RSA.value) {
 		throw new InvalidTokenTypeError();
 	}
 
 	const keyID = tokenRequest.truncatedTokenKeyId.toString();
+	const { sk, pk } = await getKeyPair(ctx, keyID);
+
+	const domain = new URL(request.url).host;
+	const issuer = new Issuer(BlindRSAMode.PSS, domain, sk, pk, { supportsRSARAW: true });
+	const signedToken = await issuer.issue(tokenRequest);
+	ctx.metrics.signedTokenTotal.inc({ key_id: keyID });
+
+	// too verbose with workers observability
+	// once there is a way to filter logpush based on log level, we can consider re-enabling
+	// console.debug(`Token issued successfully for key ${keyID}`);
+
+	return new Response(signedToken.serialize(), {
+		headers: { 'content-type': MediaType.PRIVATE_TOKEN_RESPONSE },
+	});
+};
+
+const handleBatchedTokenRequest = async (ctx: Context, request: Request): Promise<Response> => {
+	try {
+		const buffer = await request.arrayBuffer();
+		const batchedTokenRequest = BatchedTokenRequest.deserialize(new Uint8Array(buffer));
+
+		// Validate that all token requests have the same key ID and correct token type
+		const keyIds = new Set<number>();
+		for (const tokenRequest of batchedTokenRequest.tokenRequests) {
+			if (tokenRequest.tokenType !== TOKEN_TYPES.BLIND_RSA.value) {
+				throw new InvalidTokenTypeError();
+			}
+			keyIds.add(tokenRequest.truncatedTokenKeyId);
+			if (keyIds.size > 1) {
+				throw new Error("Batched token request must have the same key ID");
+			}
+		}
+
+		const firstKey = keyIds.values().next();
+		if (firstKey.done || firstKey.value === undefined) {
+			throw new Error("No valid key ID found in batched token request.");
+		}
+		const keyID = firstKey.value.toString();
+
+		const { sk, pk } = await getKeyPair(ctx, keyID);
+		const domain = new URL(request.url).host;
+
+		// We only support type 2 (Blind RSA, e.g. PSS) for now
+		const issuer = new Issuer(publicVerif.BlindRSAMode.PSS, domain, sk, pk, { supportsRSARAW: true });
+		const batchedTokenIssuer = new BatchedTokensIssuer(issuer);
+		const batchedTokenResponse = await batchedTokenIssuer.issue(batchedTokenRequest);
+
+		const responseBytes = batchedTokenResponse.serialize();
+
+		// Determine if any token response is empty (null) and choose the proper status code:
+		// If at least one token request failed, return HTTP 206 (Partial Content), otherwise 200.
+		const partial = batchedTokenResponse.tokenResponses.some((resp) => resp.tokenResponse === null);
+		const status = partial ? 206 : 200;
+
+		return new Response(responseBytes, {
+			status,
+			headers: {
+				"Content-Type": MediaType.ARBITRARY_BATCHED_TOKEN_RESPONSE,
+				"Content-Length": responseBytes.length.toString(),
+			},
+		});
+	} catch (e) {
```

Comment:
the try catch block seems to encompass more than serialisation/deserialisation. consider:
1. limiting the scope of the try catch block
2. declare a new error in [errors.ts](./errors.ts) with status code 422
3. throw the new error and let it propagate. This will allow metrics to capture that error, instead if returning a response which will not count against metrics